### ISSUE TITLE
Diffusion policy fails with async inference on a GPU

### ISSUE THREAD

**Issue by RahimMirani** [huggingface/lerobot#3861]
### Ticket Type

🐛 Bug Report (Something isn't working)

### Environment & System Info

```Shell
- LeRobot version: 0.5.x
- Platform: Linux (RunPod container, Ubuntu)
- GPU model: NVIDIA RTX A4000
- Is PyTorch built with CUDA support?: Yes
- Using GPU in script?: Yes (`policy_device=cuda`)
- Robot: SO-101 follower; 
- Client host: Raspberry Pi 5
- Cameras: scene + wrist, 640×480 MJPG
```

### The bug

Async inference crashes immediately when serving a **diffusion** policy. The policy server starts and loads the model fine, receives observation #0, then errors on the very first inference and never produces an action, the robot never moves.

An **ACT** policy runs correctly on the identical setup (same arm, cameras, server, and network path), so the crash is specific to the diffusion policy's inference path, not the async transport or the hardware.

### Error / logs

Server-side, on the first observation:

​```
INFO  y_server.py:226 Running inference for observation #0 (must_go: True)
ERROR y_server.py:266 Error in StreamActions: stack expects a non-empty TensorList
​```

After this the server just loops "Starting receiver" and no action chunks are ever generated.

### Setup / Reproduction

- **Policy server:** RunPod pod, RTX A4000, exposed TCP port
- **Robot client:** Raspberry Pi 5 (SO-101 follower), connecting over direct TCP
- **Cameras:** scene + wrist, 640×480 MJPG

Server:
​```bash
python -m lerobot.async_inference.policy_server --host=0.0.0.0 --port=8080
​```

Client (on the Pi 5):
​```bash
python -m lerobot.async_inference.robot_client \
  --server_address=<ip>:<port> \
  --robot.type=so101_follower \
  --robot.port=/dev/ttyACM1 \
  --robot.id=my_follower \
  --robot.cameras="{ scene: {type: opencv, index_or_path: /dev/video0, width: 640, height: 480, fps: 30, fourcc: MJPG}, wrist: {type: opencv, index_or_path: /dev/video2, width: 640, height: 480, fps: 30, fourcc: MJPG}}" \
  --task="any_task" \
  --policy_type=diffusion \
  --pretrained_name_or_path=username/diffusion_pick_place \
  --policy_device=cuda \
  --actions_per_chunk=100 \
  --chunk_size_threshold=0.5 \
  --aggregate_fn_name=weighted_average \
  --fps=30
​```

Swapping only `--policy_type` and `--pretrained_name_or_path` back to my ACT policy works correctly on the same command.

### Expected behavior

The diffusion policy server returns action chunks and the arm executes the task autonomously, as it does with ACT.

### Relevant logs or stack trace

```Shell

```

### Checklist

- [x] I have searched existing tickets to ensure this isn't a duplicate.
- [x] I am using the latest version of the `main` branch.
- [x] I have verified this is not an environment-specific problem.

**Comment by Lynn-hh** [comment]
I looked into this and have a minimal, hardware-free reproduction. The root
cause is that the async policy server serves observations one at a time and
calls `predict_action_chunk` directly, but Diffusion Policy requires
`n_obs_steps` (default 2) stacked observations. ACT works because its
`n_obs_steps == 1`.

Minimal repro (no robot/GPU needed):

```python
import torch
from lerobot.policies.diffusion.configuration_diffusion import DiffusionConfig
from lerobot.policies.diffusion.modeling_diffusion import DiffusionPolicy
from lerobot.configs.types import PolicyFeature, FeatureType

cfg = DiffusionConfig(
    n_obs_steps=2, horizon=16, n_action_steps=8, crop_shape=None,
    input_features={
        "observation.image": PolicyFeature(type=FeatureType.VISUAL, shape=(3, 96, 96)),
        "observation.state": PolicyFeature(type=FeatureType.STATE, shape=(4,)),
    },
    output_features={"action": PolicyFeature(type=FeatureType.ACTION, shape=(4,))},
)
stats = {k: {"mean": torch.zeros(s), "std": torch.ones(s), "min": torch.zeros(s), "max": torch.ones(s)}
         for k, s in {"observation.image": (3,1,1), "observation.state": (4,), "action": (4,)}.items()}
policy = DiffusionPolicy(cfg, dataset_stats=stats).eval()
policy.reset()
# Async server feeds a single live frame and calls predict_action_chunk directly:
policy.predict_action_chunk({
    "observation.image": torch.randn(1, 3, 96, 96),
    "observation.state": torch.randn(1, 4),
})
# -> AssertionError at modeling_diffusion.py generate_actions:
#    assert n_obs_steps == self.config.n_obs_steps
```

In the full async path this surfaces as the reporter's "stack expects a
non-empty TensorList", but the underlying cause is the same: `n_obs_steps > 1`
policies aren't given an observation history when served via
`predict_action_chunk`.

I'd be happy to open a PR. Two possible directions — which do you prefer?
1. Have `predict_action_chunk` build/maintain the observation history (mirror
   what `select_action` already does via the queues), so single-frame live
   serving fills `n_obs_steps` (copying the first obs at episode start).
2. Handle the history accumulation on the async `policy_server` side.

I'll include a regression test either way. (Note: AI-assisted investigation.)

**Comment by RahimMirani** [comment]
Opened a fix for this in #3886.

Root cause: the async `PolicyServer` calls `DiffusionPolicy.predict_action_chunk()` directly with a single, un-warmed observation and never populates the observation-history queues that `select_action()` relies on. Diffusion conditions on `n_obs_steps` past frames, so with empty queues there was no temporal observation dimension to stack, which is why it crashed on observation #0 (`stack expects a non-empty TensorList`). ACT is unaffected because it has no temporal observation queue.

The PR expands the single async frame along an `n_obs_steps` temporal dimension inside `predict_action_chunk()`, mirroring the warm-up that the synchronous `select_action()` path does via `populate_queues`.

Validated end-to-end on the original setup (RunPod RTX A4000 server, `policy_device=cuda`; Raspberry Pi 5 SO-101 client, scene + wrist cameras, async over TCP). Before the fix the server crashed at observation #0; after the fix it runs diffusion inference continuously and streams action chunks:

```
Running inference for observation #103
Preprocessing and inference took 0.9853s, action shape: torch.Size([1, 8, 6])
Action chunk #103 generated | Total time: 1020.71ms
```
